File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/even_and_odd.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

Node *evenAndOdd(Node *head) {
    if (head == nullptr || head->next == nullptr) return head;
    Node *even = head->next;
    Node *odd = head;
    Node *evenHead = head->next;

    while (even && even->next) {
        odd->next = odd->next->next;
        even->next = even->next->next;

        odd = odd->next;
        even = even->next;
    }
    odd->next = evenHead;
    return head;
}

int main() {
    vector<int> vec = {2, 1, 3, 4};
    Node *head = array_to_linkedlist(vec);
    head = evenAndOdd(head);
    traverseLinkedList(head);
    return 0;
}

File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/EvenAndOdd.java
public class EvenAndOdd {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[]{1,2,3,7});
        head = evenAndOdd(head);
        LinkedListOperations.traverseLinkedList(head);
    }

    static Node evenAndOdd(Node head) {
        if (head == null || head.next == null) return head;
        Node even = head.next;
        Node odd = head;
        Node evenHead = head.next;

        while (even != null && even.next != null) {
            odd.next = odd.next.next;
            even.next = even.next.next;

            odd = odd.next;
            even = even.next;
        }
        odd.next = evenHead;
        return head;
    }
}
